In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/hockey-players/hockey_players.csv


### Задача №1

Используя данные о хоккеистах (скачать можно здесь), проверьте, является ли среди хоккеистов из Финляндии, Норвегии и Дании значимым отличие: а) роста, б) BMI.



In [2]:
import numpy as np
import pandas as pd
from scipy import stats

In [3]:
df = pd.read_csv('../input/hockey-players/hockey_players.csv', encoding='ISO-8859-1')

In [4]:
df.tail()

year country  no            name position side  height  weight  \
6287  2016     USA  71    foligno nick        F    L     183    95.0   
6288  2016     USA  72   vatrano frank        F    L     178    98.0   
6289  2016     USA   1   kinkaid keith        G    L     191    88.0   
6290  2016     USA  30     condon mike        G    L     188    89.0   
6291  2016     USA  31  demko thatcher        G    L     193    95.0   

           birth                   club        age  cohort        bmi  
6287  1987-10-31  columbus blue jackets  28.501027    1987  28.367524  
6288  1994-03-14          boston bruins  22.132786    1994  30.930438  
6289  1989-07-04      new jersey devils  26.825462    1989  24.122146  
6290  1990-04-27     montreal canadiens  26.012320    1990  25.181077  
6291  1995-12-08         boston college  20.396988    1995  25.504040

In [5]:
def anova(*ys: np.ndarray) -> float:
    """Функция возвращает значение F-статистики
    для дисперсионного анализа.
    """
    
    y = np.concatenate(ys)

    ss_b = sum((yi.mean() - y.mean()) ** 2 * yi.size for yi in ys)
    ss_w = sum(((yi - yi.mean()) ** 2).sum() for yi in ys)
    
    sigma_b = ss_b / (len(ys) - 1)
    sigma_w = ss_w / (y.size - len(ys))
    
    return sigma_b / sigma_w

In [6]:
df['country'].unique()

array(['RUS', 'AUT', 'BLR', 'CAN', 'CZE', 'FIN', 'GER', 'ITA', 'JPN',
       'LAT', 'NOR', 'SUI', 'SVK', 'SWE', 'UKR', 'USA', 'POL', 'SLO',
       'DEN', 'FRA', 'KAZ', 'HUN'], dtype=object)

In [7]:
countries = ['FIN', 'NOR', 'DEN']
countries

['FIN', 'NOR', 'DEN']

In [8]:
ys = [df.loc[df['country'] == country, 'height'] for country in countries]

F = anova(*ys)
F

4.273207343917213

In [9]:
k = len(countries)
n = sum(y.size for y in ys)

k1 = k - 1
k2 = n - k

k1, k2

(2, 1023)

In [10]:
alpha = 0.05

t = stats.f.ppf(1 - alpha, k1, k2)
t

3.0045220661840073

Критическая область 
(
3.0045
,
+
∞
)

**Ответ:**
Поскольку статистика попала в критическую область, заключем, что между ростом хоккеистов выбранных стран есть отличия

In [11]:
bmi = [df.loc[df['country'] == country, 'bmi'] for country in countries]

F = anova(*bmi)
F

36.058167277438194

**Ответ:**
Поскольку статистика попала в критическую область, заключем, что между bmi хоккеистов выбранных стран есть отличия